iris_grib/grib_phenom_translation/_gribcode.py

```diff
@@ -0,0 +1,191 @@
+# Copyright iris-grib contributors
+#
+# This file is part of iris-grib and is released under the BSD license.
+# See LICENSE in the root of the repository for full licensing details.
+'''
+Provide object to represent grib phenomena
+
+For use as cube attributes, freely convertible to+from strings
+'''
+from dataclasses import dataclass
+import re
+from typing import Optional
+
+
+def _invalid_edition(edition):
+    msg = (
+        f"Invalid grib edition, {edition!r}, for GRIBcode : "
+        "can only be 1 or 2."
+    )
+    raise ValueError(msg)
+
+
+def _invalid_nargs(args):
+    nargs = len(args)
+    msg = (
+        f"Cannot create GRIBCode from {nargs} arguments, "
+        f"GRIBCode({args!r}) : expects either 1 or 4 arguments."
+    )
+    raise ValueError(msg)
+
+
+_RE_PARSE_FOURNUMS = re.compile(4 * r'[^\d]*(\d*)')
+
+
+def _fournums_from_gribcode_string(edcn_string):
+    parsed_ok = False
+    nums_match = _RE_PARSE_FOURNUMS.match(edcn_string).groups()
+    if nums_match is not None:
```

Comment:
BTW I **_do_** see now why this never fails.  An empty string produces 4 groups  ("", "", "", "").

The truth is, I wrote this rather a long time ago and have forgotten the principles of regex (yet again).  I guess we can rewrite it in plain code if you like.

Reply:
TODO: I'll fix this.